tensorboard/data/server/gcs/logdir.rs

```diff
@@ -0,0 +1,146 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+//! Adapter from GCS to TensorBoard logdirs.
+
+use log::warn;
+use reqwest::StatusCode;
+use std::collections::HashMap;
+use std::io::{self, BufReader, Read};
+use std::path::{Path, PathBuf};
+
+use super::Client;
+use crate::logdir::{EventFileBuf, EVENT_FILE_BASENAME_INFIX};
+use crate::types::Run;
+
+/// A reference to a GCS object with a read offset.
+pub struct File {
+    gcs: Client,
+    bucket: String,
+    object: String,
+    pos: u64,
+}
+
+impl File {
+    fn new(gcs: Client, bucket: String, object: String) -> Self {
+        Self {
+            gcs,
+            bucket,
+            object,
+            pos: 0,
+        }
+    }
+}
+
+fn reqwest_to_io_error(e: reqwest::Error) -> io::Error {
+    let kind = match e.status() {
+        Some(StatusCode::NOT_FOUND) => io::ErrorKind::NotFound,
+        Some(StatusCode::FORBIDDEN) => io::ErrorKind::PermissionDenied,
+        Some(StatusCode::UNAUTHORIZED) => io::ErrorKind::PermissionDenied,
+        Some(StatusCode::REQUEST_TIMEOUT) => io::ErrorKind::TimedOut,
+        _ if e.is_timeout() => io::ErrorKind::TimedOut,
+        _ if e.is_decode() => io::ErrorKind::InvalidData,
+        _ => io::ErrorKind::Other,
+    };
+    io::Error::new(kind, e)
+}
+
+impl Read for File {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if buf.is_empty() {
+            return Ok(0);
+        }
+        let range = self.pos..=self.pos + (buf.len() as u64 - 1);
+        let result = self
+            .gcs
+            .read(&self.bucket, &self.object, range)
+            .map_err(reqwest_to_io_error)?;
+        (&mut buf[0..result.len()]).copy_from_slice(&result);
```

Comment:
I may not be reading this right but it looks like we have 2 copies here - one inside `gcs::Client` where we do `body.to_vec()` to go from `Bytes` to `Vec<u8>` and then another one here where we go from `Vec<u8>` to our final `[u8]`.

I doubt it's really a big deal performance-wise but if that is indeed happening, it seems slightly nicer if we could do only one copy?  E.g. if we were to have our client directly return `Bytes` rather than `Vec<u8>` it seems like we could directly call `copy_from_slice(Bytes)` since `Bytes` is `Borrow<[u8]>`?

Reply:
Indeed! I was planning to do this as an optimization in a follow-up,
since using `Bytes` directly requires adding a dep on the `bytes` crate.
But it’s well spotted, and a perfectly appropriate use of `Bytes`.

I don’t see an easy way to zero-copy the bytes all the way from the HTTP
response into the `io::Read` output: e.g., neither `reqwest` nor `hyper`
seems to expose a `response.bytes_into(&mut buf)`.

There is a similar large copy in the `ReadBlob` RPC handler, which in
previous drafts I had replaced with an `Arc<[u8]>`, but `Bytes` might be
better there, too. (After all, `Bytes` is “`Arc<[u8]>` on steroids”.)

Reply:
Gotcha, fine by me to do it in a follow-up.

I may be misunderstanding but isn't it impossible to do a zero-copy transfer of HTTP response bytes into an `&mut [u8]` which is what `io::Read` takes?  Since the signature requires that we stuff the bytes into a mutable buffer we don't control?  Unless what you're saying is that we can actually change the underlying buffer pointed to by `buf` to literally just point to the one occupied by `Bytes`, but my understanding was that we can only mutate the contents of `buf`, not repoint it entirely.

Reply:
I agree that I don’t see a way to do it with the `reqwest`/et al. APIs,
but you could imagine that it could be possible if each of the layers of
the stack supported it. Something like:

```rust
// mod reqwest::blocking::response
impl Response {
    /// Streams some bytes from the response body directly into the
    /// given buffer. Returns the number of bytes written.
    fn bytes_into<T: BufMut>(&self, x: &mut T) -> crate::Result<usize> {
    }
}

// mod gcs::client
impl Client {
    fn read(&self, bucket, object, buf: impl BufMut) -> self::Result<usize> {
        let res = self.http.get(...).send()?;
        res.bytes_into(&mut buf)
    }
}

// mod gcs::logdir
impl Read for File {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let n = self.gcs.read(bucket, obj, buf)?;
        self.pos += n as u64;
        Ok(n)
    }
}
```

Or, from a lower-level perspective: at the end of the day, we receive
network data by calling `recvfrom(2)` and passing the kernel a mutable
pointer to a user-space buffer into which to write the received data, so
there’s no fundamental reason that that pointer can’t point into the
buffer given to `io::Read::read` (at least, as far as I can tell).

> but my understanding was that we can only mutate the contents of `buf`, not repoint it entirely.

Correct, and that is indeed not what I am alluding to.